packages/flutter/test/cupertino/text_field_test.dart

```diff
@@ -5289,6 +5289,73 @@ void main() {
     // The selection doesn't move beyond the left handle. There's always at
     // least 1 char selected.
     expect(controller.selection.extentOffset, 5);
+  });
+
+  testWidgets('Can drag one handle past the other on Apple platforms', (WidgetTester tester) async {
+    final TextEditingController controller = TextEditingController(
+      text: 'abc def ghi',
+    );
+    addTearDown(controller.dispose);
+    // On iOS/iPadOS, during a tap we select the edge of the word closest to the tap.
+    // On macOS, we select the precise position of the tap.
+    final bool isTargetPlatformIOS = defaultTargetPlatform == TargetPlatform.iOS;
+
+    await tester.pumpWidget(
+      CupertinoApp(
+        home: Center(
+          child: CupertinoTextField(
+            dragStartBehavior: DragStartBehavior.down,
+            controller: controller,
+            style: const TextStyle(fontSize: 10.0),
+          ),
+        ),
+      ),
+    );
+
+    // Double tap on 'e' to select 'def'.
+    final Offset ePos = textOffsetToPosition(tester, 5);
+    await tester.tapAt(ePos, pointer: 7);
+    await tester.pump(const Duration(milliseconds: 50));
+    expect(controller.selection.isCollapsed, isTrue);
+    expect(controller.selection.baseOffset, isTargetPlatformIOS ? 7 : 5);
+    await tester.tapAt(ePos, pointer: 7);
+    await tester.pumpAndSettle();
+    expect(controller.selection.baseOffset, 4);
+    expect(controller.selection.extentOffset, 7);
+
+    final RenderEditable renderEditable = findRenderEditable(tester);
+    final List<TextSelectionPoint> endpoints = globalize(
+      renderEditable.getEndpointsForSelection(controller.selection),
+      renderEditable,
+    );
+    expect(endpoints.length, 2);
+
+    // On Mac, the toolbar blocks the drag on the right handle, so hide it.
+    final EditableTextState editableTextState = tester.state(find.byType(EditableText));
+    editableTextState.hideToolbar(false);
+    await tester.pumpAndSettle();
+
+    // Drag the right handle until there's only 1 char selected.
+    // We use a small offset because the endpoint is on the very corner
+    // of the handle.
+    final Offset handlePos = endpoints[1].point;
+    Offset newHandlePos = textOffsetToPosition(tester, 5); // Position of 'e'.
+    final TestGesture gesture = await tester.startGesture(handlePos, pointer: 7);
+    await tester.pump();
+    await gesture.moveTo(newHandlePos);
+    await tester.pump();
+    expect(controller.selection.baseOffset, 4);
+    expect(controller.selection.extentOffset, 5);
+
+    newHandlePos = textOffsetToPosition(tester, 2); // Position of 'c'.
+    await gesture.moveTo(newHandlePos);
+    await tester.pump();
+    await gesture.up();
+    await tester.pump();
+
+    // The selection inverts moving beyond the left handle.
+    expect(controller.selection.baseOffset, 4);
+    expect(controller.selection.extentOffset, 2);
   }, variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.iOS }));
```

Comment:
And macOS too?

Reply:
I set it to run on iOS only since:
* There are no touch devices running macOS afaik, and selection handles can only be dragged with a touch pointer device.
* https://github.com/flutter/flutter/blob/74669e4bf1352a5134ad68398a6bf7fac0a6473b/packages/flutter/lib/src/cupertino/text_field.dart#L1121-L1127 This case triggers when we drag selection handles on macOS and immediately causes the handles to hide when dragging them, causing the test to fail after the handle is dragged initially since the handle disappears.

Reply:
I think we need to revisit how handles behave on desktop devices, there is a related issue open https://github.com/flutter/flutter/issues/108668. By default desktop platforms do not render selection handles right now, their `TextSelectionControls` explicitly block this. Right now you can provide a `TextSelectionControls` to `TextField` that allows the selection handles to render on desktop platforms, but using a mouse to manipulate them does not work since they only except touch pointer events. The case pointed out in https://github.com/flutter/flutter/pull/157815#discussion_r1870444296 also prevents selection handles from working properly on desktop platforms since they disappear immediately.

Reply:
Ok got it. Similarly I think we messed up mouse input on mobile devices for awhile, but it should work just like it does natively on those devices. So macOS is no big deal, but we should solve https://github.com/flutter/flutter/issues/108668 for Windows and Linux.